File: latest/workshop1_m3/W01_M03_demo.py
"""
============================================================
Workshop 1 — Module 3: Visual Perception & Pre-Attentive Attributes
Python Demonstration Script
Data Visualization for Data Scientists — UNYT Tirana
============================================================
"""

import numpy as np
import matplotlib.pyplot as plt
import os

os.makedirs("output", exist_ok=True)
plt.rcParams["figure.dpi"] = 150
np.random.seed(42)


# ── 1. Pre-Attentive Pop-Out: Color ────────────────────────
n = 80
xs, ys = np.random.uniform(0, 10, n), np.random.uniform(0, 8, n)

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 4))

# No pop-out: uniform color
ax1.scatter(xs, ys, s=100, c="#1565C0", edgecolors="white", linewidth=0.5)
ax1.set_title("No Pop-Out\n(serial search required)", fontsize=10, fontweight="bold")
for sp in ax1.spines.values(): sp.set_visible(False)
ax1.set_xticks([]); ax1.set_yticks([])

# Pop-out: one red among blue
colors = ["#1565C0"] * n
colors[37] = "#E53935"
sizes = [100] * n
sizes[37] = 200
ax2.scatter(xs, ys, s=sizes, c=colors, edgecolors="white", linewidth=0.5)
ax2.set_title("Color Pop-Out\n(pre-attentive: <200 ms)", fontsize=10, fontweight="bold")
for sp in ax2.spines.values(): sp.set_visible(False)
ax2.set_xticks([]); ax2.set_yticks([])

fig.suptitle("Pre-Attentive Processing: Color Channel", fontsize=12, fontweight="bold")
plt.tight_layout()
plt.savefig("output/preattentive_popout_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 2. Conjunction Search ──────────────────────────────────
n_c = 60
xc, yc = np.random.uniform(0.5, 9.5, n_c), np.random.uniform(0.5, 7.5, n_c)

fig, ax = plt.subplots(figsize=(6, 5))
for i in range(n_c):
    if i == 33:  # target: red square
        ax.scatter(xc[i], yc[i], s=160, c="#E53935", marker="s",
                   edgecolors="white", linewidth=0.5, zorder=5)
    elif i % 3 == 0:  # red circles (distractors)
        ax.scatter(xc[i], yc[i], s=100, c="#E53935", marker="o",
                   edgecolors="white", linewidth=0.5)
    elif i % 3 == 1:  # blue squares (distractors)
        ax.scatter(xc[i], yc[i], s=100, c="#1565C0", marker="s",
                   edgecolors="white", linewidth=0.5)
    else:  # blue circles
        ax.scatter(xc[i], yc[i], s=100, c="#1565C0", marker="o",
                   edgecolors="white", linewidth=0.5)

ax.set_title("Conjunction Search: Red Square Among Mixed Distractors\n"
             "No pop-out — serial scan at ~50 ms/item",
             fontsize=10, fontweight="bold")
for sp in ax.spines.values(): sp.set_visible(False)
ax.set_xticks([]); ax.set_yticks([])
plt.tight_layout()
plt.savefig("output/conjunction_search_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 3. Gestalt: Proximity via Subplots ─────────────────────
np.random.seed(7)
groups = {"A": (np.random.normal(0, 1, 50), np.random.normal(0, 1, 50)),
          "B": (np.random.normal(0, 1, 50), np.random.normal(0, 1, 50)),
          "C": (np.random.normal(0, 1, 50), np.random.normal(0, 1, 50))}

fig, axes = plt.subplots(1, 3, figsize=(10, 3.5), sharey=True)
for ax, (name, (gx, gy)) in zip(axes, groups.items()):
    ax.scatter(gx, gy, s=25, c="#1565C0", alpha=0.6, edgecolors="white", linewidth=0.3)
    ax.set_title(f"Group {name}", fontsize=10, fontweight="bold")
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    ax.tick_params(labelsize=7)

fig.suptitle("Gestalt Proximity: Physical Separation Creates Groups",
             fontsize=11, fontweight="bold")
plt.tight_layout()
plt.savefig("output/gestalt_proximity_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 4. Gestalt: Enclosure with axvspan ─────────────────────
np.random.seed(99)
months = np.arange(1, 13)
values = np.cumsum(np.random.normal(3, 5, 12)) + 50

fig, ax = plt.subplots(figsize=(7, 4))

# Enclosure: shaded band for Q2 (months 4-6)
ax.axvspan(4, 6, color="#E3F2FD", alpha=0.7, label="Q2 Promotion")
ax.plot(months, values, "o-", color="#1565C0", linewidth=1.5, markersize=5)
ax.text(5, max(values) + 2, "Q2 Promotion\nPeriod", ha="center",
        fontsize=8, fontweight="bold", color="#1565C0")

ax.set_xlabel("Month", fontsize=9)
ax.set_ylabel("Revenue ($K)", fontsize=9)
ax.set_title("Gestalt Enclosure: axvspan() Highlights a Period",
             fontsize=10, fontweight="bold")
ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)
ax.tick_params(labelsize=7)
plt.tight_layout()
plt.savefig("output/gestalt_enclosure_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 5. Multi-Channel Encoding ─────────────────────────────
np.random.seed(55)
n_e = 80
spend = np.random.uniform(10, 100, n_e)
roi = 0.4 * spend + np.random.normal(0, 12, n_e)
cat = np.random.choice(["A", "B", "C"], n_e)
budget = np.random.uniform(20, 200, n_e)

palette = {"A": "#1f77b4", "B": "#d62728", "C": "#2ca02c"}

fig, ax = plt.subplots(figsize=(7, 5))
for c in ["A", "B", "C"]:
    mask = cat == c
    ax.scatter(spend[mask], roi[mask], s=budget[mask] * 0.8,
               c=palette[c], alpha=0.6, edgecolors="white",
               linewidth=0.3, label=f"Cat {c}")

ax.set_xlabel("Marketing Spend ($K)", fontsize=9)
ax.set_ylabel("ROI (%)", fontsize=9)
ax.set_title("Four Encoding Channels: Position(x,y) + Color + Size",
             fontsize=10, fontweight="bold")
ax.legend(fontsize=8, title="Category", title_fontsize=9)
ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)
ax.grid(alpha=0.15)
ax.tick_params(labelsize=7)
plt.tight_layout()
plt.savefig("output/multi_channel_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 6. Channel Accuracy Comparison ────────────────────────
cats = ["A", "B", "C", "D"]
vals = [42, 45, 43, 44]

fig, axes = plt.subplots(1, 3, figsize=(12, 3.5))

# Position (most accurate)
axes[0].bar(cats, vals, color="#1565C0", width=0.5)
axes[0].set_ylim(0, 50)
axes[0].set_title("Position\n(most accurate)", fontsize=9, fontweight="bold")
axes[0].spines["top"].set_visible(False)
axes[0].spines["right"].set_visible(False)

# Area (less accurate)
for i, (c, v) in enumerate(zip(cats, vals)):
    r = np.sqrt(v / np.pi) * 0.3
    circle = plt.Circle((i, 0.5), r, color="#E53935", alpha=0.7)
    axes[1].add_patch(circle)
    axes[1].text(i, -0.2, c, ha="center", fontsize=9, fontweight="bold")
axes[1].set_xlim(-1, 4)
axes[1].set_ylim(-0.5, 1.5)
axes[1].set_aspect("equal")
axes[1].set_title("Area\n(less accurate)", fontsize=9, fontweight="bold")
for sp in axes[1].spines.values(): sp.set_visible(False)
axes[1].set_xticks([]); axes[1].set_yticks([])

# Color saturation (least accurate)
for i, (c, v) in enumerate(zip(cats, vals)):
    alpha = v / 50
    rect = plt.Rectangle((i - 0.35, 0), 0.7, 1,
                          facecolor=(0.08, 0.33, 0.73, alpha))
    axes[2].add_patch(rect)
    axes[2].text(i, -0.15, c, ha="center", fontsize=9, fontweight="bold")
axes[2].set_xlim(-0.8, 3.8)
axes[2].set_ylim(-0.3, 1.3)
axes[2].set_title("Saturation\n(least accurate)", fontsize=9, fontweight="bold")
for sp in axes[2].spines.values(): sp.set_visible(False)
axes[2].set_xticks([]); axes[2].set_yticks([])

fig.suptitle("Cleveland & McGill: Same Data, Different Accuracy\n"
             "Values: A=42, B=45, C=43, D=44",
             fontsize=11, fontweight="bold")
plt.tight_layout()
plt.savefig("output/channel_comparison_py.png", dpi=150, bbox_inches="tight")
plt.close()


print("── All M03 plots saved to output/ ──")

File: latest/week1/workshop1_m3/W01_M03_demo.R
# ============================================================
# Workshop 1 — Module 3: Visual Perception & Pre-Attentive Attributes
# R Demonstration Script
# Data Visualization for Data Scientists — UNYT Tirana
# ============================================================

# ── 0. Setup ────────────────────────────────────────────────
library(tidyverse)
library(patchwork)
dir.create("output", showWarnings = FALSE)


# ── 1. Pre-Attentive Pop-Out Demo ──────────────────────────
# Generate a scatter field where one point pops out by colour

set.seed(42)
n <- 80
df_pop <- tibble(
  x = runif(n, 0, 10),
  y = runif(n, 0, 8),
  target = c(rep("distractor", 36), "target", rep("distractor", n - 37))
)

# Colour pop-out: target is immediately visible
p_popout <- ggplot(df_pop, aes(x = x, y = y, color = target, size = target)) +
  geom_point(alpha = 0.8) +
  scale_color_manual(values = c(distractor = "#1565C0", target = "#E53935")) +
  scale_size_manual(values = c(distractor = 3, target = 5)) +
  theme_void() +
  theme(legend.position = "none") +
  labs(title = "Pre-Attentive Color Pop-Out",
       subtitle = "The red point is detected in <200 ms")

print(p_popout)
ggsave("output/preattentive_popout.png", p_popout,
       width = 6, height = 4, dpi = 150)


# ── 2. Conjunction Failure ─────────────────────────────────
# Red square among red circles + blue squares = serial search

set.seed(42)
n_conj <- 60
df_conj <- tibble(
  x = runif(n_conj, 0, 10),
  y = runif(n_conj, 0, 8),
  colour = sample(c("red", "blue"), n_conj, replace = TRUE),
  shape_type = sample(c("circle", "square"), n_conj, replace = TRUE)
)
# Force one target: red square at position 33
df_conj$colour[33] <- "red"
df_conj$shape_type[33] <- "square"

p_conj <- ggplot(df_conj, aes(x = x, y = y,
                               color = colour, shape = shape_type)) +
  geom_point(size = 3, alpha = 0.8) +
  scale_color_manual(values = c(red = "#E53935", blue = "#1565C0")) +
  scale_shape_manual(values = c(circle = 16, square = 15)) +
  theme_void() +
  theme(legend.position = "none") +
  labs(title = "Conjunction Search: No Pop-Out",
       subtitle = "Find the red square — requires serial scanning (~50 ms/item)")

print(p_conj)
ggsave("output/conjunction_search.png", p_conj,
       width = 6, height = 4, dpi = 150)


# ── 3. Gestalt: Proximity via Faceting ─────────────────────
# Faceting physically separates groups, leveraging proximity

set.seed(7)
df_gestalt <- tibble(
  x = rnorm(150),
  y = rnorm(150),
  group = rep(LETTERS[1:3], each = 50)
)

p_proximity <- ggplot(df_gestalt, aes(x = x, y = y)) +
  geom_point(color = "#1565C0", alpha = 0.6, size = 2) +
  facet_wrap(~group, ncol = 3) +
  theme_minimal(base_size = 10) +
  theme(legend.position = "none",
        strip.text = element_text(face = "bold")) +
  labs(title = "Gestalt Proximity via Faceting",
       subtitle = "Physical separation creates perceptual groups")

print(p_proximity)
ggsave("output/gestalt_faceting.png", p_proximity,
       width = 9, height = 3.5, dpi = 150)


# ── 4. Gestalt: Enclosure with annotate("rect") ───────────
# Highlight a region on a time series using a shaded band

set.seed(99)
df_ts <- tibble(
  date = seq(as.Date("2023-01-01"), by = "month", length.out = 12),
  value = cumsum(rnorm(12, 3, 5)) + 50
)

p_enclosure <- ggplot(df_ts, aes(x = date, y = value)) +
  # Enclosure: shaded band for Q2

  annotate("rect",
           xmin = as.Date("2023-04-01"), xmax = as.Date("2023-06-30"),
           ymin = -Inf, ymax = Inf,
           fill = "#E3F2FD", alpha = 0.6) +
  annotate("text",
           x = as.Date("2023-05-15"), y = max(df_ts$value) + 2,
           label = "Q2 Promotion Period", fontface = "bold",
           size = 3, color = "#1565C0") +
  geom_line(color = "#1565C0", linewidth = 1) +
  geom_point(color = "#1565C0", size = 2) +
  theme_minimal(base_size = 10) +
  labs(title = "Gestalt Enclosure: Shaded Band Highlights a Period",
       x = NULL, y = "Revenue ($K)")

print(p_enclosure)
ggsave("output/gestalt_enclosure.png", p_enclosure,
       width = 7, height = 4, dpi = 150)


# ── 5. Multi-Channel Encoding ─────────────────────────────
# x = position, y = position, color = nominal, size = quantitative

set.seed(55)
n_enc <- 80
df_enc <- tibble(
  spend = runif(n_enc, 10, 100),
  roi = 0.4 * spend + rnorm(n_enc, 0, 12),
  category = sample(c("A", "B", "C"), n_enc, replace = TRUE),
  budget = runif(n_enc, 20, 200)
)

p_encoding <- ggplot(df_enc, aes(x = spend, y = roi,
                                  color = category, size = budget)) +
  geom_point(alpha = 0.6) +
  scale_color_manual(values = c(A = "#1565C0", B = "#E53935", C = "#2E7D32")) +
  scale_size_continuous(range = c(1.5, 8)) +
  theme_minimal(base_size = 10) +
  labs(title = "Four Encoding Channels: x, y, color, size",
       x = "Marketing Spend ($K)", y = "ROI (%)",
       color = "Category", size = "Budget ($K)")

print(p_encoding)
ggsave("output/multi_channel.png", p_encoding,
       width = 7, height = 5, dpi = 150)


# ── 6. Cleveland & McGill: Channel Comparison ─────────────
# Same data encoded three ways to show accuracy degradation

vals <- c(A = 42, B = 45, C = 43, D = 44)

# Position (most accurate)
p_pos <- ggplot(tibble(cat = names(vals), val = vals),
                aes(x = cat, y = val)) +
  geom_col(fill = "#1565C0", width = 0.5) +
  coord_cartesian(ylim = c(0, 50)) +
  theme_minimal(base_size = 9) +
  labs(title = "Position (Best)", y = "Value", x = NULL)

# Area (less accurate)
p_area <- ggplot(tibble(cat = names(vals), val = vals),
                 aes(x = cat, y = 1)) +
  geom_point(aes(size = val), color = "#E53935", alpha = 0.7) +
  scale_size_area(max_size = 15) +
  theme_void(base_size = 9) +
  theme(axis.text.x = element_text()) +
  labs(title = "Area (Worse)", x = NULL)

# Color saturation (worst)
p_col <- ggplot(tibble(cat = names(vals), val = vals),
                aes(x = cat, y = 1, fill = val)) +
  geom_tile(height = 2, width = 0.8) +
  scale_fill_gradient(low = "#E3F2FD", high = "#1565C0") +
  theme_void(base_size = 9) +
  theme(axis.text.x = element_text(),
        legend.position = "none") +
  labs(title = "Saturation (Worst)", x = NULL)

p_channels <- p_pos | p_area | p_col
p_channels <- p_channels +
  plot_annotation(
    title = "Cleveland & McGill: Channel Accuracy Degrades",
    subtitle = "Values: A=42, B=45, C=43, D=44 — can you tell?"
  )

print(p_channels)
ggsave("output/channel_comparison.png", p_channels,
       width = 10, height = 4, dpi = 150)


cat("\n── All M03 plots saved to output/ ──\n")
